include/boost/uuid/detail/md5.hpp

```diff
@@ -0,0 +1,341 @@
+/*
+ * This RFC 1321 compatible MD5 implementation originated at:
+ * http://openwall.info/wiki/people/solar/software/public-domain-source-code/md5
+ *
+ * Author:
+ * Alexander Peslyak, better known as Solar Designer <solar at openwall.com>
+ *
+ * This software was written by Alexander Peslyak in 2001.  No copyright is
+ * claimed, and the software is hereby placed in the public domain.
+ * In case this attempt to disclaim copyright and place the software in the
+ * public domain is deemed null and void, then the software is
+ * Copyright (c) 2001 Alexander Peslyak and it is hereby released to the
+ * general public under the following terms:
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * There's ABSOLUTELY NO WARRANTY, express or implied.
+ *
+ */
+
+// Distributed under the Boost Software License, Version 1.0. (See
+// accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_UUID_MD5_HPP
+#define BOOST_UUID_MD5_HPP
+
+#include <boost/config.hpp>
+#include <boost/cstdint.hpp>
+#include <boost/uuid/uuid.hpp> // for version
+#include <string.h>
+
+namespace boost {
+namespace uuids {
+namespace detail {
+
+class md5
+{
+public:
+    typedef unsigned int(digest_type)[4];
+
+    md5()
+    {
+        MD5_Init(&ctx_);
+    }
+
+    void process_byte(unsigned char byte)
+    {
+        MD5_Update(&ctx_, &byte, 1);
+    }
+
+    void process_bytes(void const* buffer, uint32_t byte_count)
+    {
+        MD5_Update(&ctx_, buffer, byte_count);
+    }
+
+    void get_digest(digest_type& digest)
+    {
+        MD5_Final(reinterpret_cast<unsigned char *>(&digest[0]), &ctx_);
+    }
+
+    unsigned char get_version() const
+    {
+        // RFC 4122 Section 4.1.3
+        return uuid::version_name_based_md5;
+    }
+
+private:
+
+    /* Any 32-bit or wider unsigned integer data type will do */
+    typedef uint32_t MD5_u32plus;
+
+    typedef struct {
+        MD5_u32plus lo, hi;
+        MD5_u32plus a, b, c, d;
+        unsigned char buffer[64];
+        MD5_u32plus block[16];
+    } MD5_CTX;
+
+    /*
+     * The basic MD5 functions.
+     *
+     * F and G are optimized compared to their RFC 1321 definitions for
+     * architectures that lack an AND-NOT instruction, just like in Colin Plumb's
+     * implementation.
+     */
+    BOOST_FORCEINLINE MD5_u32plus F(MD5_u32plus x, MD5_u32plus y, MD5_u32plus z) { return ((z) ^ ((x) & ((y) ^ (z)))); }
```

Comment:
I thought you wanted to rename one or two caps letter long functions like this one to avoid clashes with macros, did you not?

Reply:
Sure I can do that too.